### PR TITLE
ci(jenkins): support release stage

### PR DESCRIPTION
**Pull request by v1v**
Closes https://github.com/elastic/apm-agent-ruby/issues/461

## Highlights
- Enable release stage to automatically publish to `rubygems` once a tag version has been created.
- Use the docker image `ruby:2.6` to build the gem in the release stage.
- Enable the ssh-agent to connect to GitHub within the docker container context.
- Git push author based on the last commit details.

### Further details
- The complexity is regarding the credentials to Rubygems and GitHub within a docker container created on the fly.
- Docker container used to help to be platform agnostic.
- Latest docker ruby image is the one used for the release, although it can be changed by setting the env variable: `RUBY_DOCKER_TAG`

### Test Cases
- I managed to reproduce a similar CI environment with a hello-world gem:
  - https://github.com/v1v/ruby-hello-world/tree/master
  - https://rubygems.org/gems/hola-v1v

image

### CODE REVIEW

**Comment by mikker** [comment]
Very nice, @v1v ! This looks exactly like it should. Do you need me to add the CI Rubygems user as an owner? They have support for 2FA which is probably a hassle to set up?

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -178,6 +179,33 @@ pipeline {
...
+          }
+        }
+        steps {
+          withGithubNotify(context: 'Release') {
+            deleteDir()
+            unstash 'source'
+            script {
+              rubygemsLogin.withApi(secret: RUBYGEM) {
+                sh 'rake release'
+                sh 'gem pull x.y'
```
Comment (mikker) [inline .ci/Jenkinsfile]: would this check that the release went through by downloading it immediately afterwards?

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -178,6 +179,35 @@ pipeline {
...
+            expression { return params.Run_As_Master_Branch }
+          }
+        }
+        steps {
+          withGithubNotify(context: 'Release') {
+            deleteDir()
+            unstash 'source'
+            script {
+              sshagent(['f6c7695a-671e-4f4f-a331-acdce44ff9ba']) {
+                rubygemsLogin.withApi(secret: RUBYGEM) {
```
Comment (mdelapenya) [inline .ci/Jenkinsfile]: Must `RUBYGEM` be quoted?

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -178,6 +179,35 @@ pipeline {
...
+            tag pattern: 'v\\d+.*', comparator: 'REGEXP'
+            expression { return params.Run_As_Master_Branch }
+          }
+        }
+        steps {
+          withGithubNotify(context: 'Release') {
+            deleteDir()
+            unstash 'source'
+            script {
+              sshagent(['f6c7695a-671e-4f4f-a331-acdce44ff9ba']) {
```
Comment (kuisathaverat) [inline .ci/Jenkinsfile]: I am not sure if the key will be in the Docker container

File: .ci/Jenkinsfile (hunk 1 above)
Comment (v1v) [reply .ci/Jenkinsfile]: It's not required, so I just removed :)

File: .ci/Jenkinsfile (hunk 3 above)
Comment (v1v) [reply .ci/Jenkinsfile]: I'd go for the docker.inside + sshagent then

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -206,8 +206,9 @@ pipeline {
                   withEnv(["HOME=${env.WORKSPACE}/${env.BASE_DIR ?: ''}"]) {
                     rubygemsLogin.withApi(secret: "${env.RELEASE_SECRET}") {
                       sshagent(['f6c7695a-671e-4f4f-a331-acdce44ff9ba']) {
-                        //sh "bundle install && rake release"
-                        sh 'ssh -T git@github.com | true'
+                        sh '.ci/prepare-git-context.sh'
+                        sh 'bundle install'
```
Comment (mikker) [inline .ci/Jenkinsfile]: I don't think we need the (rather heavy) bundle. Try without or perhaps with only `gem install rake` and then `rake release`

Comment (v1v) [reply .ci/Jenkinsfile]: - without the bundle 

I get something like the below error in my local environment
```
[Pipeline] sh
+ rake release
rake aborted!
LoadError: cannot load such file -- yard
```

- with `gem install rake`

```
+ gem install rake
Successfully installed rake-12.3.3
1 gem installed
[Pipeline] sh
+ rake release
rake aborted!
LoadError: cannot load such file -- yard
```

My local environment does use a kind of similar Gemfile:
- https://github.com/v1v/ruby-hello-world/blob/master/Gemfile

What should be do here?

Comment (mikker) [reply .ci/Jenkinsfile]: Looking inside `Rakefile`, `gem install rake yard rspec` should do it.

Comment (v1v) [reply .ci/Jenkinsfile]: Awesome, it worked like a charm

image

Comment (v1v) [reply .ci/Jenkinsfile]:
```suggestion
                        sh 'gem install rake yard rspec'
```

**Comment by v1v** [comment]
As far as I see the test error happens with the `jruby:9.2`. I don't think that's related to this particular PR.

[test error details](https://apm-ci.elastic.co/job/apm-agent-ruby/job/apm-agent-ruby-downstream/job/PR-530/81/testReport/junit/(root)/transactions/Test___Ruby_jruby_9_2_rails_5_1___Rails_integration_transactions_validates_json_schema/)


image

**Comment by estolfo** [comment]
@v1v no, that error is not related to this PR

File: .ci/Jenkinsfile [hunk 5]
```diff
@@ -178,6 +179,40 @@ pipeline {
...
+      stage('Release') {
+        options { skipDefaultCheckout() }
+        environment {
+          RUBY_DOCKER_TAG = 'ruby:2.6'
+        }
+        when {
+          beforeAgent true
+          anyOf {
+            tag pattern: 'v\\d+.*', comparator: 'REGEXP'
+            expression { return params.Run_As_Master_Branch }
```
Comment (mikker) [inline .ci/Jenkinsfile]: Use some other kind of mechanism. `Run as master` is sometimes used for benchmarking PRs
```suggestion
```

**Pull request merged**